internal/ocp/ca/helper.go

```diff
@@ -0,0 +1,199 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	clusterCAType = "cluster"
+	serviceCAType = "service"
+	typeKey       = "kmm.openshift.io/ca.type"
+)
+
+type ConfigMap struct {
+	KeyName string
+	Name    string
+}
+
+var errConfigMapNotFound = errors.New("ConfigMap not found")
+
+//go:generate mockgen -source=helper.go -package=ca -destination=mock_helper.go
+
+type Helper interface {
+	GetClusterCA(ctx context.Context, namespace string) (*ConfigMap, error)
+	GetServiceCA(ctx context.Context, namespace string) (*ConfigMap, error)
+	Sync(ctx context.Context, namespace string, owner client.Object) error
+}
+
+type helperImpl struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func NewHelper(client client.Client, scheme *runtime.Scheme) Helper {
+	return &helperImpl{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+func (h *helperImpl) Sync(ctx context.Context, namespace string, owner client.Object) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	logger.Info("Syncing the Cluster CA ConfigMap")
+
+	clusterCM, err := h.getOrCreateClusterCAConfigMap(ctx, namespace)
+	if err != nil {
+		return fmt.Errorf("could not get the cluster CA ConfigMap: %v", err)
+	}
+
+	if err = h.syncConfigMapOwners(ctx, clusterCM, owner, logger); err != nil {
+		return fmt.Errorf("could not sync the cluster CA ConfigMap owners: %v", err)
+	}
+
+	logger.Info("Syncing the service CA ConfigMap")
+
+	serviceCM, err := h.getOrCreateServiceCAConfigMap(ctx, namespace)
+	if err != nil {
+		return fmt.Errorf("could not get the service CA ConfigMap: %v", err)
+	}
+
+	if err = h.syncConfigMapOwners(ctx, serviceCM, owner, logger); err != nil {
+		return fmt.Errorf("could not sync the service CA ConfigMap owners: %v", err)
+	}
+
+	logger.Info("Synced CA ConfigMaps")
+
+	return nil
+}
+
+func (h *helperImpl) GetClusterCA(ctx context.Context, namespace string) (*ConfigMap, error) {
+	cm, err := h.getConfigMap(ctx, namespace, clusterCAType)
+	if err != nil {
+		return nil, fmt.Errorf("could not get CA ConfigMap: %v", err)
+	}
+
+	caCM := ConfigMap{
+		KeyName: "ca-bundle.crt",
+		Name:    cm.Name,
+	}
+
+	return &caCM, nil
+}
+
+func (h *helperImpl) GetServiceCA(ctx context.Context, namespace string) (*ConfigMap, error) {
+	cm, err := h.getConfigMap(ctx, namespace, serviceCAType)
+	if err != nil {
+		return nil, fmt.Errorf("could not get CA ConfigMap: %v", err)
+	}
+
+	caCM := ConfigMap{
+		KeyName: "service-ca.crt",
+		Name:    cm.Name,
+	}
+
+	return &caCM, nil
+}
+
+func (h *helperImpl) getOrCreateClusterCAConfigMap(ctx context.Context, namespace string) (*v1.ConfigMap, error) {
+	cm, err := h.getConfigMap(ctx, namespace, clusterCAType)
+	if err != nil {
+		if !errors.Is(err, errConfigMapNotFound) {
+			return nil, fmt.Errorf("could not get the cluster CA ConfigMap: %v", err)
+		}
+
+		cm = &v1.ConfigMap{
```

Comment:
why do we want to create an empty ConfigMap?

Reply:
OpenShift controllers populate the contents based on the label or annotation.

Reply:
So we can still get the scenario where the first time we run worker Pods, they may fail, since the ConfigMaps are not updated yet, right?

Reply:
We can still hit that.